### ISSUE TITLE
Moving away from relying on setuptools develop mode

### ISSUE THREAD

**Issue by bmorris3** [spacetelescope/jdaviz#1729]
Invoking `setup.py` is deprecated (you can catch up on what's preferred [in this blog post](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html), for example). Currently, jdaviz still uses the `setuptools.setup` function in `develop` mode to make symlinks for our custom voila template, named `jdaviz-default`. The template is defined in the [source code here](https://github.com/spacetelescope/jdaviz/tree/main/share/jupyter), and the linking on `python setup.py develop` is done [here](https://github.com/spacetelescope/jdaviz/blob/main/setup.py#L123-L164).

An improvement option to investigate: 
* `hatch` is a python package manager that may cover our needs. Both voila ([source](https://github.com/voila-dashboards/voila/blob/main/hatch_build.py)) and Jupyter ([source](https://github.com/jupyterlab/hatch-jupyter-builder), [blog](https://blog.jupyter.org/packaging-for-jupyter-in-2022-c7be64c38926)) recently migrated to hatch, with the help of their [hatch-jupyter-builder](https://hatch-jupyter-builder.readthedocs.io/en/latest/source/how_to_guides/migrating_data_files_projects.html
) package.

Aside: @rosteen and I have found it helpful to find the available (=="installed"?) templates by running a script [like this](https://gist.github.com/bmorris3/fecb8b6bda475ddcdf34df7b337f2c75). Maybe we should offer a util like this to voila upstream?

**Comment by pllim** [comment]
Original motivation: @rosteen reported that he got missing template error running `jdaviz` CLI but only if `pip install -e .` was done and not `pip install .`, which would suggest that our current way of doing development install no longer works. I asked him to test this hypothesis to see if https://setuptools.pypa.io/en/latest/userguide/development_mode.html#legacy-behavior would fix the development install (which would confirm that indeed our stuff is legacy). Ricky needs to get back on this.

> voila upstream?

That would be a question for @maartenbreddels .

**Comment by ofek** [comment]
Hatchling works by default for `-e`, see https://github.com/microsoft/pylance-release/issues/3407

**Comment by maartenbreddels** [comment]
Hi @bmorris3 

I think it would be best to first get a working version of say hatch and what is in https://github.com/voila-dashboards/voila/pull/1250 in jdaviz before looking at https://github.com/voila-dashboards/voila/pull/1250.
It may be a good idea to try to get `pip install -e .` working first.
I'm also not sure how pep-660 changes the situation, and of hatchling already supports that.

cheers,

Maarten

**Comment by ofek** [comment]
> It may be a good idea to try to get `pip install -e .` working first. I'm also not sure how pep-660 changes the situation, and of hatchling already supports that.

Yup supported! `pip` introduced support for [PEP 660](https://peps.python.org/pep-0660/) in [v21-3](https://pip.pypa.io/en/stable/news/#v21-3). Also, with Hatchling:

- the implementation doesn't pollute the project with a `*.egg-info` directory
- supports static analysis tools for editable installations by default whereas the new setuptools [does not](https://github.com/microsoft/pylance-release/issues/3407) i.e. it requires enabling an option for IDEs to work

**Comment by maartenbreddels** [comment]
Could you point us to documentation/code on how this works/how this is implemented?

**Comment by ofek** [comment]
https://hatch.pypa.io/latest/config/build/#dev-mode

Just `.pth` files unless enabling `dev-mode-exact`

**Comment by stscijgbot-jwql** [comment]
_This issue is tracked on JIRA as [JDAT-2842](https://jira.stsci.edu/browse/JDAT-2842)._

**Comment by stscijgbot-jwql** [comment]
_Comment by [Brett Morris](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bmmorris) on [JIRA](https://jira.stsci.edu/browse/JDAT-2842?focusedCommentId=614616#comment-614616):_

Also made an upstream PR to voilà, in case this effort belongs there: voila-dashboards/voila#1250
